src/audio/pipeline.c

```diff
@@ -378,11 +378,62 @@ static int pipeline_comp_hw_params(struct comp_dev *current,
 	return 0;
 }
 
+static int pipeline_comp_params_neg(struct comp_dev *current,
+				    struct comp_buffer *calling_buf,
+				    struct pipeline_walk_context *ctx,
+				    int dir)
+{
+	struct pipeline_data *ppl_data = ctx->comp_data;
+	uint32_t flags = 0;
+	int err = 0;
+
+	pipe_dbg(ppl_data->p, "pipeline_comp_params_neg(), current->comp.id = %u, dir = %u",
+		 dev_comp_id(current), dir);
+
+	/* return if current is already configured */
+	if (current->state != COMP_STATE_INIT &&
+	    current->state != COMP_STATE_READY) {
+		if (buffer_params_match(calling_buf,
+					&ppl_data->params->params))
+			return 0;
+		/*
+		 * the param is conflict with an active pipeline,
+		 * drop an error and reject the .params() command.
+		 */
+		pipe_err(ppl_data->p, "pipeline_comp_params_neg(): params conflict with existed active pipeline!");
+		return -EINVAL;
+	}
+
+	/*
+	 * Negotiation only happen when the current component has > 1
+	 * source or sink, we are propagating the params to branched
+	 * buffers, and the subsequent component's .params() or .prepare()
+	 * should be responsible for calibrating if needed. For example,
+	 * a component who has different channels input/output buffers
+	 * should explicitly configure the channels of the branched buffers.
+	 */
+	if (calling_buf) {
```

Comment:
can we call this "connected_buf". it makes more sense.

Reply:
a component can have several source_buf and several sink_buf, they are all connected_buf? the calling_buf actually means "the propagating upstream buffer" here.

Reply:
Then lets call it "upstream_buffer"

Reply:
This "calling_buf" is already used though the whole file, @lgirdwood if you have suggestion about changing its naming, can we make it another separated PR?

Reply:
ok, a subsequent PR should change this name to also include the "origin" of the buffer. i.le. it's either coming from the upstream data source or downstream data sink.